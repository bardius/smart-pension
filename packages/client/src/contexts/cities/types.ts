import type { Dispatch, SetStateAction } from 'react'
import type { FetchResult } from '@apollo/client'
import type { MutationFunctionOptions } from '@apollo/client/react/types/types'
import type { ApolloError } from '@apollo/client/errors'
import type { City } from '../../domain/cities/types'
import type { CitiesFilters, CitiesMutationInput } from '../../domain/cities/types'

export type CitiesProviderState = {
  citiesList?: City[]
  loadingCitiesList?: boolean
  errorCitiesList?: ApolloError
  updatedCity?: City | null
  loadingUpdateCity?: boolean
  errorUpdateCity?: ApolloError
  updateCity: (
    options?: MutationFunctionOptions<{ updateCity: City }, { input: CitiesMutationInput }>
  ) => Promise<FetchResult<{ updateCity: City }>>
  filter?: Partial<CitiesFilters>
  setFilter?: Dispatch<SetStateAction<Partial<CitiesFilters>>>
  clientSideFilter?: string
  setClientSideFilter?: Dispatch<SetStateAction<string>>
}
