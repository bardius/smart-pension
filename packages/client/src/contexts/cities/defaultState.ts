import type { CitiesProviderState } from './types'

const defaultProviderState: CitiesProviderState = {
  updateCity: (): any => null,
}

export { defaultProviderState }
