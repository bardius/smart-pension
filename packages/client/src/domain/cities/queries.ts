import { gql } from '@apollo/client'

const GET_CITIES = gql`
  query GetCities($filter: CitiesFilters, $limit: Int, $offset: Int) {
    cities(filter: $filter, limit: $limit, offset: $offset) {
      cities {
        id
        name
        country
        visited
        wishlist
      }
      total
    }
  }
`

const UPDATE_CITY = gql`
  mutation UpdateCity($input: CitiesMutationInput) {
    updateCity(input: $input) {
      id
      visited
      wishlist
    }
  }
`

export { GET_CITIES, UPDATE_CITY }
