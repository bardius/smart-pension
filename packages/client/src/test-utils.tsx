import * as React from 'react'
import type { RenderOptions } from '@testing-library/react'
import { render } from '@testing-library/react'
import { ChakraProvider, theme } from '@chakra-ui/react'
import { MockedProvider } from '@apollo/client/testing'
import { citiesQueriesMocks } from './__mocks__/citiesQueries'

const AllProviders = ({ children }: { children?: React.ReactNode }) => (
  <MockedProvider mocks={citiesQueriesMocks} addTypename={false}>
    <ChakraProvider theme={theme}>{children}</ChakraProvider>
  </MockedProvider>
)

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllProviders, ...options })

export { customRender as render }
