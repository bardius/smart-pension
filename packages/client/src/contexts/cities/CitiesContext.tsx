import { createContext } from 'use-context-selector'
import { defaultProviderState } from './defaultState'
import type { CitiesProviderState } from './types'

const CitiesContext = createContext<CitiesProviderState>(defaultProviderState)

export { CitiesContext }
