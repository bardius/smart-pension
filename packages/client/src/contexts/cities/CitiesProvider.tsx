import { useEffect, useMemo, useState } from 'react'
import type { FC } from 'react'
import { useMutation, useQuery } from '@apollo/client'
import { GET_CITIES, UPDATE_CITY } from '../../domain/cities/queries'
import type { City, CitiesFilters, CitiesMutationInput, CitiesResult } from '../../domain/cities/types'
import { sortCitiesData } from '../../domain/cities/service'
import { CitiesContext } from './CitiesContext'

// Holds all the logic of the state management and the required request orchestration
// and any mutation or actions of the datasets like sorting (added a sample sorting)
const CitiesProvider: FC = ({ children }) => {
  const [filter, setFilter] = useState<Partial<CitiesFilters>>({})
  const [clientSideFilter, setClientSideFilter] = useState<string>('')

  // Define queries/mutations
  const {
    loading: loadingCitiesList,
    error: errorCitiesList,
    data: dataCitiesList,
    refetch: refetchCitiesList,
  } = useQuery<CitiesResult, { filter: Partial<CitiesFilters> }>(GET_CITIES, {
    variables: { filter: filter },
  })

  // Instead of re-fetching all cities data we could update the cached city list
  // after mutation with update option and cache.modify
  // We can also force refresh list via options with { refetchQueries: [GET_CITIES, 'GetCities'] }
  const [updateCity, { loading: loadingUpdateCity, error: errorUpdateCity, data: dataUpdateCity }] = useMutation<
    { updateCity: City },
    { input: CitiesMutationInput }
  >(UPDATE_CITY)

  // Sample sorting
  // for serverside pagination the sorting should also be done serverside and not here
  const sortedCityList = useMemo(() => {
    return sortCitiesData(dataCitiesList)
  }, [dataCitiesList])

  // One way of implementing this if we do not wish to get all cities in one go
  // and then just do the filtering on the UI is to use useLazyQuery and then
  // keep updating the filters in state this will cause new request per typed
  // char. Leaving as only performance improvement the debounce on change or
  // to change the event handler on blur for the search input
  useEffect(() => {
    refetchCitiesList({ filter: filter })
  }, [refetchCitiesList, filter])

  // Expose values via the provider state
  const stateValues = {
    citiesList: sortedCityList,
    loadingCitiesList: loadingCitiesList,
    errorCitiesList: errorCitiesList,
    updatedCity: dataUpdateCity?.updateCity,
    loadingUpdateCity: loadingUpdateCity,
    errorUpdateCity: errorUpdateCity,
    setFilter: setFilter,
    clientSideFilter: clientSideFilter,
    setClientSideFilter: setClientSideFilter,
    updateCity: updateCity,
  }

  return <CitiesContext.Provider value={stateValues}>{children}</CitiesContext.Provider>
}

export { CitiesProvider }
