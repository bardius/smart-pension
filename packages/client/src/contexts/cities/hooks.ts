import { useContext, useContextSelector } from 'use-context-selector'
import type { ApolloError } from '@apollo/client/errors'
import type { Dispatch, SetStateAction } from 'react'
import type { MutationFunctionOptions } from '@apollo/client/react/types/types'
import type { FetchResult } from '@apollo/client'
import { filterObjectArrayByBooleanKeyValue, filterObjectArrayByPartialKeyValue } from '../../domain/cities/service'
import type { City } from '../../domain/cities/types'
import type { CitiesFilters, CitiesMutationInput } from '../../domain/cities/types'
import type { CitiesProviderState } from './types'
import { CitiesContext } from './CitiesContext'

const useCitiesState = (): CitiesProviderState => {
  const context = useContext(CitiesContext)

  if (context === undefined) {
    throw new Error('CitiesContext must be used within CitiesProvider')
  }

  return context
}

const useCitiesList = (): City[] | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.citiesList)

  return contextSlice
}

const useCitiesListLoading = (): boolean | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.loadingCitiesList)

  return contextSlice
}

const useCitiesListError = (): ApolloError | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.errorCitiesList)

  return contextSlice
}

const useCitiesFilter = (): Partial<CitiesFilters> | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.filter)

  return contextSlice
}

const useWishlistCitiesList = (): City[] | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.citiesList)

  return contextSlice ? filterObjectArrayByBooleanKeyValue<City>(contextSlice, 'wishlist', true) : contextSlice
}

const useVisitedCitiesList = (): City[] | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.citiesList)

  return contextSlice ? filterObjectArrayByBooleanKeyValue<City>(contextSlice, 'visited', true) : contextSlice
}

const useCitiesSetFilter = (): Dispatch<SetStateAction<Partial<CitiesFilters>>> | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.setFilter)

  return contextSlice
}

const useCitiesSetClientSideFilter = (): Dispatch<SetStateAction<string>> | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.setClientSideFilter)

  return contextSlice
}

const useFilteredCitiesList = (): City[] | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.citiesList)
  const clientSideFilter = useContextSelector(CitiesContext, state => state.clientSideFilter)

  return contextSlice ? filterObjectArrayByPartialKeyValue<City>(contextSlice, 'name', clientSideFilter) : contextSlice
}

const useUpdatedCity = (): City | null | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.updatedCity)

  return contextSlice
}

const useUpdateCityLoading = (): boolean | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.loadingUpdateCity)

  return contextSlice
}

const useUpdateCityError = (): ApolloError | undefined => {
  const contextSlice = useContextSelector(CitiesContext, state => state.errorUpdateCity)

  return contextSlice
}

const useUpdateCity = (): ((
  options?: MutationFunctionOptions<{ updateCity: City }, { input: CitiesMutationInput }>
) => Promise<FetchResult<{ updateCity: City }>>) => {
  const contextSlice = useContextSelector(CitiesContext, state => state.updateCity)

  return contextSlice
}

export {
  useCitiesState,
  useCitiesList,
  useCitiesListLoading,
  useCitiesListError,
  useCitiesFilter,
  useWishlistCitiesList,
  useVisitedCitiesList,
  useCitiesSetFilter,
  useCitiesSetClientSideFilter,
  useFilteredCitiesList,
  useUpdatedCity,
  useUpdateCityLoading,
  useUpdateCityError,
  useUpdateCity,
}
