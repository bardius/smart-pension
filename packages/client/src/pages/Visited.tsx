import React from 'react'
import type { VFC } from 'react'
import { Container, Heading, VStack } from '@chakra-ui/react'
import { useVisitedCitiesList } from '../contexts/cities/hooks'
import { CityList } from '../components/CityList'

export const Visited: VFC = () => {
  const citiesList = useVisitedCitiesList()

  return (
    <VStack as="main" spacing="8" data-testid="visited-page">
      <Heading as="h1">Visited</Heading>
      <Container centerContent maxW="container.md" flexDir="column">
        <CityList cities={citiesList} />
      </Container>
    </VStack>
  )
}
