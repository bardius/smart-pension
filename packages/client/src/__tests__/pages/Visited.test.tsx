import React from 'react'
import { screen } from '@testing-library/react'
import { render } from '../../test-utils'
import { Visited } from '../../pages/Visited'
import { CitiesProvider } from '../../contexts/cities/CitiesProvider'

describe('Page - Visited', () => {
  it('should render the Visited page content', async () => {
    const renderLayout = render(
      <CitiesProvider>
        <Visited />
      </CitiesProvider>
    )

    expect(screen.getByTestId('visited-page')).toBeInTheDocument()
    expect(await screen.findByTestId('city-list')).toBeInTheDocument()

    const { asFragment } = renderLayout
    expect(asFragment()).toMatchSnapshot()
  })
})
