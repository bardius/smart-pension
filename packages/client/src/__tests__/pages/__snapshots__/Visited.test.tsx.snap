// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page - Visited should render the Visited page content 1`] = `
<DocumentFragment>
  <main
    class="chakra-stack css-uc6l60"
    data-testid="visited-page"
  >
    <h1
      class="chakra-heading css-1dklj6k"
    >
      Visited
    </h1>
    <div
      class="chakra-container css-ed19g4"
    >
      <ul
        class="css-khhkvr"
        data-testid="city-list"
      >
        <li
          class="css-16ax18g"
          data-testid="city-card"
        >
          <div
            class="css-qiuk2s"
          >
            <p
              class="chakra-text css-wjomy9"
            >
              Athens
            </p>
            <span
              class="chakra-text css-3b0vjl"
            >
              (Greece)
            </span>
          </div>
          <div
            class="css-zdpt2t"
          >
            <button
              aria-label="Marked as visited, Click to toggle status"
              class="chakra-button css-8x7hex"
              data-testid="city-card-visited-btn-active"
              title="Marked as visited, Click to toggle status"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M23.555,8.729a1.505,1.505,0,0,0-1.406-.98H16.062a.5.5,0,0,1-.472-.334L13.405,1.222a1.5,1.5,0,0,0-2.81,0l-.005.016L8.41,7.415a.5.5,0,0,1-.471.334H1.85A1.5,1.5,0,0,0,.887,10.4l5.184,4.3a.5.5,0,0,1,.155.543L4.048,21.774a1.5,1.5,0,0,0,2.31,1.684l5.346-3.92a.5.5,0,0,1,.591,0l5.344,3.919a1.5,1.5,0,0,0,2.312-1.683l-2.178-6.535a.5.5,0,0,1,.155-.543l5.194-4.306A1.5,1.5,0,0,0,23.555,8.729Z"
                  fill="currentColor"
                />
              </svg>
            </button>
            <button
              aria-label="Added in wishlist, Click to toggle status"
              class="chakra-button css-8x7hex"
              data-testid="city-card-wishlist-btn-active"
              title="Added in wishlist, Click to toggle status"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon css-onkibi"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <g
                  fill="none"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-width="2"
                >
                  <rect
                    height="18"
                    rx="2"
                    ry="2"
                    width="18"
                    x="3"
                    y="3"
                  />
                  <path
                    d="M12 8v8"
                  />
                  <path
                    d="M8 12h8"
                  />
                </g>
              </svg>
            </button>
          </div>
        </li>
      </ul>
    </div>
  </main>
</DocumentFragment>
`;
